Transport: allow to de-serialize arbitrary objects given their name

javanna opened the pull request:
This commit makes it possible to serialize arbitrary objects by having them extend Writeable. When reading them though, we need to be able to identify which object we have to create, based on its name. This is useful for queries once we move to parsing on the coordinating node, as well as with aggregations and so on.

Introduced a new abstraction called NamedWriteable, which is supported by StreamOutput and StreamInput through writeNamedWriteable and readNamedWriteable methods. A new NamedWriteableRegistry is introduced also where named writeable prototypes need to be registered so that we are able to retrieve the proper instance of the writeable given its name and then de-serialize it calling readFrom against it.

We decided to streamline the support for NamedWriteables and make related methods available across the board in StreamInput and StreamOutput. That said the new write* and read* methods are package private so they can be tested but won't be made public. The idea is to add specific methods once we have named writeable to be streamed, e.g.:

```
public QueryBuilder readQuery() {
    return readNamedWriteable("query");
}
```

and

```
public void writeQuery(QueryBuilder queryBuilder) {
    writeNamedWriteable("query", queryBuilder);
}
```

The above methods cannot be added yet as neither queries nor aggs are streamable yet.

colings86 commented:
Not sure how it could be done without producing a massive PR but its a shame that anything using the NamedWriteableAwareStreamInput needs to cast the StreamInput it receives.

Also I wonder if we should have a NamedWriteableAwareStreamOutput so we are sure that classes using NamedWritable are sure to write the object properly. So it would have writeNamedWritable(NamedWritable), writeOptionalNamedWritable(NamedWritable) and write NamedWritableArray(NamedWritable[])?

It seems like this change is going to get complicated due to the mechanism for determining when to wrap the StreamInput so I agree that maybe we should explore how big the context argument option is?

javanna commented:
> Also I wonder if we should have a NamedWriteableAwareStreamOutput so we are sure that classes using NamedWritable are sure to write the object properly. So it would have writeNamedWritable(NamedWritable), writeOptionalNamedWritable(NamedWritable) and write NamedWritableArray(NamedWritable[])?

That is why we have the new serializer object that exposes the methods to read and write named writeables, you have to effectively go through it so you can read and write named writeables. That said I agree with you the casting is not great, and the current wrapping of the stream is even worse :)

I am all for adding a context argument to all readFrom methods at this point. The only condition is that the context needs to expose final objects only and must not change its state while reading.

Let's see what @jpountz and @s1monw think about this.

javanna commented:
After talking to @s1monw and @jpountz we decided to go back to something closer to the original implementation (what we have in the query-refactoring branch). We wrap the stream (only in case of request) and named writeables are supported across the board. The default registry is empty if the stream is not wrapped with one that has a non empty registry. Also we went for exposing specific readQuery and readAggregation method in the future rather than the generic readNamedWriteable and writeNamedWriteable methods. I updated the description of the PR accordingly and removed the WIP label, this is ready for review now.

jpountz commented on core/src/main/java/org/elasticsearch/common/io/stream/FilterStreamInput.java:
```diff
@@ -0,0 +1,72 @@
+
+    @Override
+    public StreamInput setVersion(Version version) {
+        return delegate.setVersion(version);
```
the builder pattern is trappy here: I think this should call delegate.setVersion() and then return this? I quickly looked and I don't see many call sites relying on the fact that StreamInput.setVersion returns itself, so maybe we can make it void to remove the trap?

javanna replied:
indeed good catch...

jpountz commented on core/src/main/java/org/elasticsearch/common/io/stream/NamedWriteableRegistry.java:
```diff
@@ -0,0 +1,63 @@
+
+    private final Map<String, NamedWriteable<?>> registry = new HashMap<>();
+
+    public static final NamedWriteableRegistry EMPTY = new NamedWriteableRegistry();
```
this singleton is a bit dangerous given that NamedWriteableRegistry is mutable, I think we should just create a new one every time?

javanna replied:
I see, but we never register stuff while using... I wonder if there is a way to seal the registry, I guess there isn't?

jpountz commented on core/src/main/java/org/elasticsearch/common/io/stream/NamedWriteableRegistry.java:
```diff
@@ -0,0 +1,63 @@
+    }
+
+    private String getKey(String category, String name) {
+        return category + "_" + name;
```
I'm not too happy with this serialization, since category=a and name =b_c would get the same key as category=a_c and name = c. Can we just have a `Map<String, Map<String, NamedWriteable<?>>` instead, where the first level is the category and the second one is the name?

javanna replied:
I see, I was trying to avoid the map in a map with this basic impl, also thought about a different separator, but I can go with your suggestion, sure

jpountz commented on core/src/main/java/org/elasticsearch/common/io/stream/NamedWriteableRegistry.java:
```diff
@@ -0,0 +1,63 @@
+        if (namedWriteable == null) {
+            throw new IllegalArgumentException("unknown named writeable with name [" + name + "] and category [" + category + "]");
+        }
+        return namedWriteable;
```
Can we avoid the unchecked cast and just return a NamedWriteable<?>

javanna replied:
somewhere we need to cast though, would you be happier if the cast was moved to StreamInput#readNamedWriteable ?

jpountz replied:
We need to cast indeed, but I want to give the compiler opportunities to find errors, which is never possible when one starts definiting methods whose generic parameter is only used in the return value. By the way I'm thinking that we could make casts more safe by making category a class instead of a string, and this class would be the base class of the object that the namedwriteables can deserialize

javanna replied:
agreed, this reminds me a lot of the very first prototype that I came up with, will work on this.

jpountz commented on core/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java:
```diff
@@ -556,6 +562,42 @@ public Object readGenericValue() throws IOException {
+    <C> C readNamedWriteable(String category) throws IOException {
+        String name = readString();
+        NamedWriteable<C> namedWriteable = namedWriteableRegistry.getPrototype(category, name);
+        return namedWriteable.readFrom(this);
```
Should we make this method throw an exception instead and have the ability to provide a namedWriteableRegistry in a subclass of FilterStreamInput which would first try to read using its own registry and otherwise delegate to its parent stream input?

javanna replied:
not sure I follow why that would be useful, can you explain? Isn't the default empty registry a reasonable default already? it is going to throw exception anyway if the registry is not available or if the object cannot be read. Maybe I am missing the point though :)

jpountz replied:
It's not a big deal, but I liked that it allows to
 - make StreamInput not depend on NamedWriteableRegistry
 - stack registries: you could stack registers on top on each other, while the current approach requires that you provide a registry that knows about all namedwriteables at once. I don't have use-cases in mind, but thought this could be useful.

javanna replied:
I will give this a shot. What feels a bit weird is that StreamInput would have methods to support NamedWriteables but it really doesn't support them as it doesn't have a registry, only a subclass will. Will play around with this though and see what it looks like after the change.

javanna commented on core/src/main/java/org/elasticsearch/common/io/stream/FilterStreamInput.java:
```diff
@@ -0,0 +1,80 @@
+/**
+ * Wraps a {@link StreamInput} and associates it with a {@link NamedWriteableRegistry}
+ */
+public class FilterStreamInput extends StreamInput {
```
I am now wondering if naming is still fine, maybe NamedWriteableAwareStreamInput or something along those lines (and shorter!) would be better?

jpountz replied:
We could keep this FilterStreamInput that just delegates everything, and add a new NamedWriteableAwareStreamInput (or just WriteableStreamInput) that extends it and adds the logic to look into the registry to deserialize?

javanna replied:
good point, will do

javanna commented on core/src/main/java/org/elasticsearch/common/io/stream/NamedWriteableRegistry.java:
```diff
@@ -0,0 +1,84 @@
+     */
+    public synchronized <T> void registerPrototype(Class<T> categoryClass, NamedWriteable<? extends T> namedWriteable) {
+        @SuppressWarnings("unchecked")
+        InnerRegistry<T> innerRegistry = (InnerRegistry<T>)registry.get(categoryClass.getName());
```
shall we check that the categoryClass is the same as the categoryClass stored in the inner registry? Looking for ways make this even safer

jpountz replied:
hmm, why do you use the class name instead of the class object itself as a key?

javanna replied:
not sure why, can change that :)

javanna commented:
@jpountz I updated the PR coording to your comments, it's ready for another round of review

jpountz commented on core/src/main/java/org/elasticsearch/common/io/stream/FilterStreamInput.java:
```diff
@@ -0,0 +1,80 @@
+    }
+
+    @Override
+    <C> C readNamedWriteable(@SuppressWarnings("unused") Class<C> categoryClass) throws IOException {
```
I think it should be:
`<C> C readNamedWriteable(@SuppressWarnings("unused") Class<? extends C> categoryClass) throws IOException {`

jpountz replied:
also what is categoryClass tagged as unused? I see it used to look up the registry?

javanna replied:
the annotation is a bug, it is needed only in the base class, will fix. I am not sure about the Class<? extends C>. If we are deserializing an arbitrary query, the base class QueryBuilder implements NamedWriteable, but when we read an arbitrary query we cannot expect a subclass of it, we just do `QueryBuilder query = readQuery();` which will call `readNamedWriteable(QueryBuilder.class)`. That is why I kept Class<C> as a type. Does it make sense?

jpountz commented on core/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java:
```diff
@@ -571,4 +600,10 @@ public static StreamInput wrap(byte[] bytes, int offset, int length) {
+    /**
+     * Wraps a {@link StreamInput} and enhances it so that it knows how to deserialize {@link NamedWriteable} arbitrary objects.
+     */
+    public static StreamInput wrap(StreamInput streamInput, NamedWriteableRegistry namedWriteableRegistry) {
```
I would put it on the namedwriteablestreaminput rather than here? Or maybe we don't actually need this method, this would just be the constructor?

javanna replied:
yea we don't really need it can remove

jpountz commented on core/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java:
```diff
@@ -556,6 +549,42 @@ public Object readGenericValue() throws IOException {
+            list.add(obj);
+        }
+        return list;
+    }
```
These methods are pkg-private, but I think they need to be public for consumption by other components?

jpountz replied:
Oh, you want to keep them pkg-private on purpose and then have the actual methods delegate to them and be public?

javanna replied:
yep that was the idea, it is confusing cause they are only used for testing.... prod code that uses them will come later.

jpountz commented on core/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java:
```diff
@@ -556,6 +549,42 @@ public Object readGenericValue() throws IOException {
+     * Reads an optional {@link NamedWriteable} from the current stream, by first reading its name and then looking for
+     * the corresponding entry in the registry by name, so that the proper object can be read and returned.
+     */
+    final <C> C readOptionalNamedWriteable(Class<C> categoryClass) throws IOException {
```
s/Class<C>/Class<? extends C>/

jpountz commented on core/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java:
```diff
@@ -556,6 +549,42 @@ public Object readGenericValue() throws IOException {
+     * Default implementation throws {@link UnsupportedOperationException} as StreamInput doesn't hold a registry.
+     * Use {@link FilterInputStream} instead which wraps a stream and supports a {@link NamedWriteableRegistry} too.
+     */
+    <C> C readNamedWriteable(@SuppressWarnings("unused") Class<C> categoryClass) throws IOException {
```
`s/Class<C>/Class<? extends C>/`

javanna replied:
I think it makes sense as-is for the reason stated above

jpountz commented on core/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java:
```diff
@@ -556,6 +549,42 @@ public Object readGenericValue() throws IOException {
+     * Reads a list of {@link NamedWriteable} from the current stream, by first reading its size and then
+     * reading the individual objects using {@link #readNamedWriteable(Class)}.
+     */
+    final <C> List<C> readNamedWriteableList(Class<C> categoryClass) throws IOException {
```
`s/Class<C>/Class<? extends C>/`

javanna replied:
I think it makes sense as-is for the reason stated above

jpountz commented on core/src/main/java/org/elasticsearch/common/io/stream/StreamOutput.java:
```diff
@@ -613,4 +613,35 @@ public void writeThrowable(Throwable throwable) throws IOException {
+        for (NamedWriteable obj : list) {
+            writeNamedWriteable(obj);
+        }
+    }
```
pkg-private -> public?

javanna commented:
I pushed another commit that should address the last review

jpountz commented:
LGTM
